[WC-494] Accordion web - animations and icons

IIsaku opened the pull request:
## Checklist
- Contains unit tests ✅ 
- Contains breaking changes ❌
- Contains Atlas changes ✅ 
- Compatible with: MX 9️⃣

#### Web specific
- Contains e2e tests ❌
- Is accessible ✅ 
- Compatible with Studio ✅ 
- Cross-browser compatible ✅ 

#### Feature specific
- Comply with designs ✅ 
- Comply with PM's requirements ✅ 

## This PR contains
- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
Introduce animations, animation properties and icon properties.

## Relevant changes
- Add animation properties for the header icon and content box.
- Fix hover text and icon colours.
- Add default icon collapse/expand animation.
- Add default content box collapse/expand animation.
- Add properties to configure custom icons and to position/disable them.
- Add advanced mode toggle.
- Align header font styles with PM requirements.

## What should be covered while testing?
Test the new properties and their conditional visibility. Also test the animations on different browsers. Test the other properties as well, since some major refactoring has been done with this PR.

IIsaku committed:
feat(accordion-web): add animating chevron to header

IIsaku committed:
feat(accordion-web): change chevron color on hover

IIsaku committed:
feat: improve accordion color styles

- Apply black color to chevron & text to non hovered header in functional styles.
- Apply (hover) text color to chevron in Atlas.

IIsaku committed:
feat(accordion-web): add content box collapse animation

- Remove lazy mounting as this doesn't work well with animations (calculating the target height).
- The collapsing animation needs some fine tuning.

IIsaku committed:
feat(accordion-web): add icon & animation properties

IIsaku committed:
feat(accordion-web): showIcon

- Display icon left, right or not inside the header.
- Apply specific classes for left and right icon.
- Don't render icon when showIcon is set to "no".
- Update unit tests to check showIcon classes and no icon rendering.
- Rename unit tests due to removal of lazy loading.
- Improve chevron animation.

IIsaku committed:
refactor(accordion-web): refs & onTransitionEnd

IIsaku committed:
feat(accordion-web): don't render icon when not collapsible

IIsaku committed:
feat(accordion-web): conditional icon properties

- If show icon is set to no, all other icon properties are hidden.

IIsaku committed:
feat(accordion-web): animate icon setting

IIsaku committed:
refactor(accordion-web): move show header icon logic to AccordionGroup

- Improve unit tests

IIsaku committed:
feat(accordion-web): custom icons

- If animateIcon is yes, allow only one icon to be configured

IIsaku committed:
fix(accordion-web): custom icon switching

IIsaku committed:
feat(accordion-web): animate collapsing

IIsaku committed:
feat(accordion-web): improve icon accessibility

IIsaku committed:
refactor(accordion-web): prop improvements

IIsaku committed:
fix(accordion-web): initiate icon collapsing animation immediately

IIsaku committed:
fix(accordion-web): collapse animation

- It's necessary to wait a little so that the browser
can settle with the newly set collapsing class.

IIsaku committed:
feat(accordion-web): change header font styles

IIsaku committed:
refactor(accordion-web): collapse animation timeout

IIsaku committed:
feat: improve accordion web styles

IIsaku committed:
refactor(accordion-web): component properties & state

- Convert group property AccordionGroup into separate properties.
- Simplify Accordion state and its reducer.
- Fix unnecessary rerenders.

IIsaku committed:
chore(accordion-web): rename reducer file

IIsaku committed:
refactor(accordion-web): extract useIconGenerator to file

IIsaku committed:
feat(accordion-web): improve styles

- Add 50 ms to all animations for the animations to start
at around the same time when in single expanded mode.
- Apply flex 1 to header content.

IIsaku commented:
I am working on fixing all unit tests.

IIsaku committed:
chore(accordion-web): rename reducer test file

IIsaku committed:
test(accordion-web): fix reducer tests

diego-antonelli commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -1,35 +1,43 @@
 
 export default function Accordion(props: AccordionProps): ReactElement | null {
-    const previousGroupsPropValue = useRef(props.groups);
+    const reducer = useRef(getCollapsedAccordionGroupsReducer(props.singleExpandedGroup ? "single" : "multiple")); // the accordion group reducer function doesn't need to change during the lifetime of this component, since the singleExpandedGroup won't change.
```
Why do we need to pass a ref to the reducer?

IIsaku replied:
The `getCollapsedAccordionGroupsReducer` will be called on every rerender when called inside the useReducer function. Making use of a `useCallback` hook doesn't work either, because the `singleExpandedGroup` prop get loads of updates due to the new objects that are being created in the root component.

diego-antonelli commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -1,35 +1,43 @@
-    const [accordionGroups, accordionGroupsDispatch] = useReducer(
-        getAccordionGroupsReducer(props.collapsible ? (props.singleExpandedGroup ? "single" : "multiple") : "all"), // the accordion group reducer function doesn't need to change during the lifetime of this component, since the singleExpandedGroup won't change.
-        props.groups.map(group => ({ ...group, collapsed: props.collapsible }))
+    const [collapsedAccordionGroups, collapsedAccordionGroupsDispatch] = useReducer(
```
NOTE: I have the feeling the reducer is not needed at all based on our chat this morning about the data change while using conditional visible tabs.

IIsaku replied:
I think the reducer is still useful because it hides the complexity of updating state logic with regard to the different expand modes. It's also a nice abstraction to be shared by both web and native widget in the end.

IIsaku committed:
test(accordion-web): fix AccordionGroup existing tests

IIsaku committed:
test(accordion-web): accordion group generateIcon & dynamic class

IIsaku committed:
test(accordion-web): accordion group

IIsaku committed:
test(accordion-web): icon

IIsaku committed:
feat(atlas-ui): add border radius to accordion web section

IIsaku committed:
test(accordion-web): components/accordion

diego-antonelli commented on packages/pluggableWidgets/accordion-web/src/Accordion.tsx:
```diff
@@ -1,12 +1,12 @@
 import { createElement, ReactElement, useMemo } from "react";
 
-import AccordionComponent from "./components/Accordion";
+import AccordionComponent, { AccordionGroups } from "./components/Accordion";
```
Seems a bit weird the imports as you have more than one export.. also commented in Accordion.tsx

IIsaku replied:
Changed it.

diego-antonelli reviewed (changes requested):
Good PR! Just nit pick tiny details.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -1,35 +1,43 @@
+            {...group}
+            collapsed={collapsedAccordionGroups[index]}
+            animateCollapsing={props.animateCollapsing}
+            generateIcon={props.generateHeaderIcon}
```
Yes, although I do think this is the best solution in this case. The alternative would be to pass AccordionGroup components as children to `src/components/Accordion`. The downside is however that we need to use context to let the `src/components/Accordion` component control the collapsed states of the groups.

diego-antonelli replied:
Lets leave like this for now

IIsaku committed:
chore(accordion-web): process review feedback

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -1,35 +1,43 @@
+    showGroupHeaderIcon?: "right" | "left" | "no";
 }
 
 export default function Accordion(props: AccordionProps): ReactElement | null {
```
Changed it.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -1,65 +1,119 @@
-        if (group.visible !== previousVisiblePropValue.current || !group.collapsed) {
-            previousVisiblePropValue.current = group.visible;
-            setDivContentMounted(group.visible && !group.collapsed);
+    const rootElement = useRef<HTMLDivElement>(null);
```
As discussed we keep it with the refs.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -1,65 +1,119 @@
+
+                setTimeout(() => {
+                    if (contentWrapperElement.current && contentElement.current) {
+                        contentWrapperElement.current.style.height = `${
```
Refactored it.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -1,65 +1,119 @@
+                    }
+                }, 50);
+            } else {
+                rootElement.current.classList.add("widget-accordion-group-expanding");
```
I would keep this as is, because it keeps the code more readable I believe.

diego-antonelli replied:
[DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) is a nice way to start

IIsaku replied:
I understand that you don't want to repeat yourself, but lines 62 & 63 are about starting an animation and lines 33 & 34 are about stopping an animation. There is nothing in common here to abstract away. Every line in a generic function will include a ternary if statement to set the right class. I don't think that's an improvement. I am wondering how you would call the generic function that combines this logic?

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -1,65 +1,119 @@
             >
-                {group.header}
+                {props.header}
+                {props.toggleCollapsed && showHeaderIcon !== "no" && showHeaderIcon
```
Changed it to `props.toggleCollapsed && (showHeaderIcon === "left" || showHeaderIcon === "right")`.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/Icon.tsx:
```diff
@@ -0,0 +1,59 @@
+                className={classNames("widget-accordion-group-header-icon", {
+                    "widget-accordion-group-header-icon-animate": animate
+                })}
+                aria-hidden={"true"}
```
Done.

diego-antonelli replied:
It is still here

diego-antonelli replied:
Btw, in line 40 and 53 you even removed the ="true" so maybe we can apply same pattern here

IIsaku commented on packages/pluggableWidgets/accordion-web/src/utils/CollapsedAccordionGroupsReducer.ts:
```diff
@@ -0,0 +1,15 @@
+export type CollapsedAccordionGroupsReducerAction = { type: "expand" | "collapse"; index: number };
+
+export function getCollapsedAccordionGroupsReducer(
```
Done.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/Icon.tsx:
```diff
@@ -0,0 +1,59 @@
+                })}
+                src={data.iconUrl}
+                alt=""
+                aria-hidden="true"
```
Good one! I will include this note for the accessibility story.

IIsaku committed:
refactor(accordion-web): useIconGenerator custom hook

- Extract dynamic value logic to the root Accordion component,
so that this hook can be used by design mode as well in the future.

IIsaku committed:
test(accordion-web): useIconGenerator custom hook

IIsaku committed:
chore(accordion-web): add hooks testing lib dep

diego-antonelli commented on packages/pluggableWidgets/accordion-web/package.json:
```diff
@@ -27,7 +27,8 @@
   "devDependencies": {
-    "@mendix/pluggable-widgets-tools": ">=9.0.0"
+    "@mendix/pluggable-widgets-tools": ">=9.0.0",
+    "@testing-library/react-hooks": "^3.4.2"
```
Not necessary, its being included already indirectly via lerna dependencies.

diego-antonelli commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -1,65 +1,119 @@
     );
 }
+
+function setContentWrapperHeightToContentHeight(contentWrapper: HTMLDivElement, content: HTMLDivElement): void {
```
You can be more specific in this function.. like receive the target height and the target element and rename the method to `setHeight`

IIsaku committed:
chore(accordion-web): process review feedback

IIsaku committed:
fix: accordion-web styles & content height changes behavior

- Prevent onTransitionEnd handler to be called when content height changes.
- Fix nested accordion styles.

IIsaku committed:
chore(accordion-web): improve property description

- Convert CollapseBehavior to ExpandBehavior property.
- Rename options,
so that options combined text length isn't exceeding Studio's limit.
- Update icon snapshot
